Python/ceval.c

```diff
@@ -4066,7 +4066,88 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
             CHECK_EVAL_BREAKER();
             DISPATCH();
         }
-
+        case TARGET(CALL_FUNCTION_ADAPTIVE): {
+            SpecializedCacheEntry *cache = GET_CACHE();
+            if (cache->adaptive.counter == 0) {
+                next_instr--;
+                if (_Py_Specialize_CallFunction(BUILTINS(), stack_pointer,
+                    cache->adaptive.original_oparg, next_instr, cache) < 0) {
+                    goto error;
+                }
+                DISPATCH();
+            }
+            else {
+                STAT_INC(CALL_FUNCTION, deferred);
+                cache->adaptive.counter--;
+                oparg = cache->adaptive.original_oparg;
+                JUMP_TO_INSTRUCTION(CALL_FUNCTION);
+            }
+        }
+        case TARGET(CALL_CFUNCTION_FAST): {
+            /* Builtin METH_FASTCALL functions, without keywords */
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            _PyCallCFunctionCache *cache1 = &caches[-1].call_function;
+            PyObject **callable = &PEEK(cache0->original_oparg + 1);
+            DEOPT_IF(!PyCFunction_CheckExact(*callable), CALL_FUNCTION);
+            PyCFunction cfunc = PyCFunction_GET_FUNCTION(*callable);
+            DEOPT_IF(cfunc != cache1->cfunc, CALL_FUNCTION);
+
+            assert(cframe.use_tracing == 0);
+            // res = func(self, args, nargs)
+            PyObject *res = ((_PyCFunctionFast)(void(*)(void))cfunc)(
+                PyCFunction_GET_SELF(*callable),
+                stack_pointer - cache0->original_oparg,
+                cache0->original_oparg);
+            assert((res != NULL) ^ (_PyErr_Occurred(tstate) != NULL));
+
+            /* Clear the stack of the function object. */
+            while (stack_pointer > callable) {
+                PyObject *x = EXT_POP(stack_pointer);
+                Py_DECREF(x);
+            }
+            PUSH(res);
+            if (res == NULL) {
+                /* Not deopting because this doesn't mean our optimization was wrong.
+                   `res` can be NULL for valid reasons. Eg. getattr(x, 'invalid').
+                   In those cases an exception is set, so we must handle it.
+                */
+                goto error;
+            }
+            record_cache_hit(cache0);
+            STAT_INC(CALL_FUNCTION, hit);
+            DISPATCH();
+        }
+        case TARGET(CALL_CFUNCTION_O): {
+            /* Builtin METH_O functions */
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            _PyCallCFunctionCache *cache1 = &caches[-1].call_function;
+            PyObject **callable = &PEEK(cache0->original_oparg + 1);
```

Comment:
Check the `oparg` has the correct value, i.e. 1, when specializing. So, this becomes `PyObject *callable = SECOND()`
Also, why the extra indirection on the pointer?

Reply:
The extra indirection was required solely for the while loop clearing the function object after the call. It seems that the address of ``callable`` changes after the function call. So passing ``&callable`` to the while loop didn't work and would cause the entire stack to be cleared.

To make this more readable, I removed the indirection and stored the initial address in another variable, like what ``call_function`` currently does.